### PR TITLE
Add build-concurrency flag

### PR DESCRIPTION
**Pull request by sladyn98**
Fixes: https://github.com/GoogleContainerTools/skaffold/issues/5647

**Description**

So this change adds the build concurrency flag to run builds in parallel, instead of having to replace it in every file we would like to provide it as a build option
**User facing changes (remove if N/A)**

**Follow-up Work (remove if N/A)**

### CODE REVIEW

File: pkg/skaffold/runner/runcontext/context.go [hunk 1]
```diff
@@ -185,6 +185,7 @@ func (rc *RunContext) Tail() bool                                { return rc.Opt
 func (rc *RunContext) Trigger() string                           { return rc.Opts.Trigger }
 func (rc *RunContext) WaitForDeletions() config.WaitForDeletions { return rc.Opts.WaitForDeletions }
 func (rc *RunContext) WatchPollInterval() int                    { return rc.Opts.WatchPollInterval }
+func (rc *RunContext) BuildConcurrency() int                     { return rc.Opts.BuildConcurrency }
```
Comment (sladyn98) [inline pkg/skaffold/runner/runcontext/context.go:188]: Would I need to call this from builder mux https://github.com/GoogleContainerTools/skaffold/blob/cc262b48867959858e592f76674f8ea0f37e0d8e/pkg/skaffold/build/builder_mux.go#L34

File: pkg/skaffold/config/options.go [hunk 2]
```diff
@@ -85,6 +85,7 @@ type SkaffoldOptions struct {
 	Command            string
 	RPCPort            int
 	RPCHTTPPort        int
+	BuildConcurrency   int
```
Comment (sladyn98) [inline pkg/skaffold/config/options.go:88]: I tried to follow the conventions of other flags hence added it here

File: cmd/skaffold/app/cmd/flags.go [hunk 3]
```diff
@@ -509,6 +509,14 @@ var flagRegistry = []Flag{
 		DefinedOn:     []string{"debug", "dev", "run"},
 		IsEnum:        true,
 	},
+	{
+		Name:          "build-concurrency",
+		Usage:         "Number of concurrently running builds",
+		Value:         &opts.AssumeYes,
```
Comment (gsquared94) [inline cmd/skaffold/app/cmd/flags.go:515]:
```suggestion
		Value:         &opts.BuildConcurrency,
```

File: cmd/skaffold/app/cmd/flags.go [hunk 4]
```diff
@@ -509,6 +509,14 @@ var flagRegistry = []Flag{
...
+		DefValue:      1,
```
Comment (gsquared94) [inline cmd/skaffold/app/cmd/flags.go:516]: we should only use the value in this flag if it was explicitly set by the user. Only way could be to have the default be `-1` instead, and have a check later if it was overridden by a non-negative integer.

File: pkg/skaffold/runner/runcontext/context.go (hunk 1 above)
Comment (gsquared94) [reply pkg/skaffold/runner/runcontext/context.go:188]: In `builder_mux.go` you can add this method to the `Config` interface:
```go

type Config interface {
	GetPipelines() []latest.Pipeline
	DefaultRepo() *string
	GlobalConfig() string
	BuildConcurrency() int  
}
```

In `NewBuilderMux` function, skip the `minConcurrency` [calculations](https://github.com/GoogleContainerTools/skaffold/blob/e947844e558b9181911906399cb5e3df7fdb28ef/pkg/skaffold/build/builder_mux.go#L62-L73) if the value from that method is non-negative (see above comment also).

File: pkg/skaffold/config/options.go (hunk 2 above)
Comment (gsquared94) [reply pkg/skaffold/config/options.go:88]: looks good

File: cmd/skaffold/app/cmd/flags.go (hunk 4 above)
Comment (sladyn98) [reply cmd/skaffold/app/cmd/flags.go:516]: Yeah this sounds goo

**Comment by aaron-prindle** [comment]
It might also make sense to update the docs w/ this flag:
docs/content/en/docs/references/cli/_index.md

See #4039 for an example

**Comment by sladyn98** [comment]
Thanks @aaron-prindle I ran the script

**Comment by gsquared94** [comment]
Please run `make linters` and fix whatever is offending it. Everything else looks good!

File: pkg/skaffold/build/builder_mux.go [hunk 5]
```diff
@@ -59,17 +60,22 @@ func NewBuilderMux(cfg Config, store ArtifactStore, builder func(p latest.Pipeli
...
-		case minConcurrency < 0:
-			minConcurrency = concurrency
-			logrus.Infof("build concurrency first set to %d parsed from %s[%d]", minConcurrency, reflect.TypeOf(b).String(), i)
-		case concurrency > 0 && (minConcurrency == 0 || concurrency < minConcurrency):
-			minConcurrency = concurrency
-			logrus.Infof("build concurrency updated to %d parsed from %s[%d]", minConcurrency, reflect.TypeOf(b).String(), i)
-		default:
-			logrus.Infof("build concurrency value %d parsed from %s[%d] is ignored since it's not less than previously set value %d", concurrency, reflect.TypeOf(b).String(), i, minConcurrency)
+
+		if cfg.BuildConcurrency() > 0 {
```
Comment (gsquared94) [inline pkg/skaffold/build/builder_mux.go:64]:
```suggestion
		if cfg.BuildConcurrency() >= 0 {
```

concurrency can be set to `0` which means run all in parallel.

File: cmd/skaffold/app/cmd/flags.go [hunk 6]
```diff
@@ -509,6 +509,14 @@ var flagRegistry = []Flag{
 		DefinedOn:     []string{"debug", "dev", "run"},
 		IsEnum:        true,
 	},
+	{
+		Name:          "build-concurrency",
+		Usage:         "Number of concurrently running builds",
```
Comment (gsquared94) [inline cmd/skaffold/app/cmd/flags.go:514]: Can you be a bit more verbose:
 `Number of concurrently running builds. Set to 0 to run all builds in parallel. Doesn't violate build order among dependencies.`

Comment (gsquared94) [reply cmd/skaffold/app/cmd/flags.go:514]: Sorry for the bits-n-pieces review.

Comment (gsquared94) [reply cmd/skaffold/app/cmd/flags.go:514]: will have to run `./hack/generate-man.sh` script again to update the docs

**Review by gsquared94 — APPROVED** [review]
lgtm!

**Pull request merged**